src/IB/IBFESurfaceMethod.cpp

```diff
@@ -1406,111 +2110,944 @@ IBFESurfaceMethod::initializeLevelData(Pointer<BasePatchHierarchy<NDIM> > hierar
+        qrule.reset();
+        unsigned int qp_offset = 0;
+        for (unsigned int e_idx = 0; e_idx < num_active_patch_elems; ++e_idx)
+        {
+            Elem* const elem = patch_elems[e_idx];
+            for (unsigned int d = 0; d < NDIM; ++d)
+            {
+                X_dof_map.dof_indices(elem, X_dof_indices[d], d);
+            }
+            get_values_for_interpolation(x_node, *X_ghost_vec, X_dof_indices);
+
+            P_jump_dof_map_cache.dof_indices(elem, P_jump_dof_indices);
+            get_values_for_interpolation(P_jump_node, *P_jump_ghost_vec, P_jump_dof_indices);
+
+            const bool qrule_changed =
+                FEDataManager::updateInterpQuadratureRule(qrule, d_default_interp_spec, elem, x_node, patch_dx_min);
+
+            if (qrule_changed)
+            {
+                fe->attach_quadrature_rule(qrule.get());
+            }
+            fe->reinit(elem);
+
+            const unsigned int n_node = elem->n_nodes();
+            const unsigned int n_qp = qrule->n_points();
+
+            // Zero out the values of X, du, and dv prior to accumulation.
+            double* x_begin = &x_qp[NDIM * qp_offset];
+            std::fill(x_begin, x_begin + NDIM * n_qp, 0.0);
+
+            double* x_io_begin = &x_io_qp[NDIM * qp_offset];
+            std::fill(x_io_begin, x_io_begin + NDIM * n_qp, 0.0);
+
+            double* N_begin = &N_qp[NDIM * qp_offset];
+            std::fill(N_begin, N_begin + NDIM * n_qp, 0.0);
+
+            double* P_jump_begin = &P_jump_qp[qp_offset];
+            std::fill(P_jump_begin, P_jump_begin + n_qp, 0.0);
+
+            // Interpolate X, du, and dv at all of the quadrature points
+            // via accumulation, i.e., X(qp) = sum_k X_k * phi_k(qp) for
+            // each qp.
+
+            for (unsigned int qp = 0; qp < n_qp; ++qp)
+            {
+                for (unsigned int k = 0; k < NDIM - 1; ++k)
+                {
+                    interpolate(dx_dxi[k], qp, x_node, *dphi_dxi[k]);
+                }
+                if (NDIM == 2)
+                {
+                    dx_dxi[1] = VectorValue<double>(0.0, 0.0, 1.0);
+                }
+                n = (dx_dxi[0].cross(dx_dxi[1])).unit();
+
+                for (unsigned int i = 0; i < NDIM; ++i)
+                {
+                    for (unsigned int k = 0; k < n_node; ++k)
+                    {
+                        const double& p_X = phi[k][qp];
+                        x_io_qp[NDIM * (qp_offset + qp) + i] += x_node[k][i] * p_X;
+                        x_qp[NDIM * (qp_offset + qp) + i] += x_node[k][i] * p_X;
+                    }
+                    N_qp[NDIM * (qp_offset + qp) + i] = n(i);
+                    // Note that here we calculate the pressure on one side as the jump plus the pressure on the other
+                    // side
+                    x_io_qp[NDIM * (qp_offset + qp) + i] += (d_calculate_interior_side ? 1.0 : -1.0) * n(i) * dh;
```

Comment:
Suggest making the flag about interior/exterior side an argument to the function instead of a class member variable.